samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,340 @@
-    // source, config, and context are used to initialize the window kv-store
-    window = spec.getWindow();
+  // The trigger state corresponding to each {@link TriggerKey}.
+  private final Map<TriggerKey<WK>, TriggerImplWrapper> triggers = new HashMap<>();
+  private final Clock clock;
+
+  public WindowOperatorImpl(WindowOperatorSpec<M, WK, WV> spec, Clock clock) {
+    this.clock = clock;
+    this.window = spec.getWindow();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    WindowKey<WK> storeKey =  getStoreKey(message);
+    WindowState<WV> existingState = store.get(storeKey);
+    WindowState<WV> newState = applyFoldFunction(existingState, message);
+
+    store.put(storeKey, newState);
+
+    if (window.getEarlyTrigger() != null) {
+      TriggerKey<WK> triggerKey = new TriggerKey<>(FiringType.EARLY, storeKey);
+
+      getOrCreateTriggerImplWrapper(triggerKey, window.getEarlyTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+
+    if (window.getDefaultTrigger() != null) {
+      TriggerKey<WK> triggerKey = new TriggerKey<>(FiringType.DEFAULT, storeKey);
+      getOrCreateTriggerImplWrapper(triggerKey, window.getDefaultTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long now = clock.currentTimeMillis();
+    TriggerCallbackState state;
+    while ((state = pendingCallbacks.peek()) != null && state.getScheduledTimeMs() <= now) {
+      pendingCallbacks.remove();
+      state.getCallback().run();
+
+      TriggerImplWrapper triggerImplWrapper = triggers.get(state.triggerKey);
+      if (triggerImplWrapper != null) {
+        triggerImplWrapper.onTimer(collector, coordinator);
+      }
+    }
+  }
 
+  /**
+   * Get the key to be used for lookups in the store for this message.
+   */
+  private WindowKey<WK> getStoreKey(M message) {
+    Function<M, WK> keyExtractor = window.getKeyExtractor();
+    WK key = null;
+
+    if (keyExtractor != null) {
+      key = keyExtractor.apply(message);
+    }
+
+    String paneId = null;
+
+    if (window.getWindowType() == WindowType.TUMBLING) {
+      long triggerDurationMs = ((TimeTrigger<M>) window.getDefaultTrigger()).getDuration().toMillis();
+      final long now = clock.currentTimeMillis();
+      Long windowBoundary = now - now % triggerDurationMs;
+      paneId = windowBoundary.toString();
+    }
+
+    return new WindowKey<>(key, paneId);
+  }
+
+  private WindowState<WV> applyFoldFunction(WindowState<WV> existingState, M message) {
+    WV wv;
+    long earliesttimestamp;
+
+    if (existingState == null) {
+      wv = window.getInitializer().get();
+      earliesttimestamp = clock.currentTimeMillis();
+    } else {
+      wv = existingState.getWindowValue();
+      earliesttimestamp = existingState.getEarliestRecvTime();
```

Comment:
Naming: `existingState.getEarliestTimestamp()` for consistency.

Naming: `earliestTimestamp` (camelcase)